### ISSUE TITLE
GPU acceleration for scikit-learn via H2O4GPU 

### ISSUE THREAD

**Issue by MattiaVerticchio** [pycaret/pycaret#304]
`pycaret` is great, however, there’s no GPU support. I’m using it Google Colab and it can be very slow with large datasets. Can I suggest using [H2O4GPU](https://github.com/h2oai/h2o4gpu) as a replacement for scikit-learn? It’s available as a [Python package](https://pypi.org/project/h2o4gpu/).

> H2O4GPU is a collection of GPU solvers by H2Oai with APIs in Python and R. The Python API builds upon the easy-to-use scikit-learn API and its well-tested CPU-based algorithms. It can be used as a drop-in replacement for scikit-learn (i.e. `import h2o4gpu as sklearn`) with support for GPUs on selected (and ever-growing) algorithms. H2O4GPU inherits all the existing scikit-learn algorithms and falls back to CPU algorithms when the GPU algorithm does not support an important existing scikit-learn class option.

**Comment by Yard1** [comment]
h2oai seems to have issues with python > 3.6, doesn't it?

**Comment by pycaret** [comment]
@MattiaVerticchio Can you have a look at code on dev-1.0.1 branch? Maybe you would like to work on including this?

**Comment by MattiaVerticchio** [comment]
> h2oai seems to have issues with python > 3.6, doesn't it?

Support for Python >3.6 is been worked on [right now](https://github.com/h2oai/h2o4gpu/pull/842).





> @MattiaVerticchio Can you have a look at code on dev-1.0.1 branch? Maybe you would like to work on including this?

Sure, I’ll  replace `sklearn` with `h2o4gpu` and test it on the tutorial examples. 👍

**Comment by MattiaVerticchio** [comment]
I swapped `sklearn` with `h2o4gpu` [for supported calls](https://github.com/pycaret/pycaret/compare/dev-1.0.1...MattiaVerticchio:dev-1.0.1) and tried to `compare_models()`. There’s an overall 20% reduction in comparing times and `GradientBoostingClassifier` had a drop of ~90% in training time when running the Beginner Classification tutorial.

In my opinion, it could be a good improvement in training times, especially for large datasets, but it should be furtherly tested. I had an issue with measuring some error metrics with the `LogisticRegression` model when running `compare_results()`.

You can test it in [this Colab Notebook](https://colab.research.google.com/drive/1w6OvzPhKkEWyaJ5_WT8HgfdGnv1w0Dad?usp=sharing
).

**Comment by pycaret** [comment]
@MattiaVerticchio It looks great. I can't understand why metrics for Logistic Regression is messed up. I agree with the change but we must handle it through setup function. That's where all the global variables are declared. We will create a new parameter "use_gpu" in setup function and then declare it global and make important dependent on the global variable. I will write a detailed user story in this week.

**Comment by MattiaVerticchio** [comment]
> In the mean time, if you can troubleshoot what's going on with LR?

As of now, I figured out the AUC metric issue only. H2O4GPU was using a couple of outdated default parameters for `LinearRegression`. I opened two issues in `h2o4gpu` repo to fix [`solver`](https://github.com/h2oai/h2o4gpu/issues/856#issue-673297145) and [`multi_class`](https://github.com/h2oai/h2o4gpu/issues/857#issue-673305775).

**Comment by webpolis** [comment]
Indeed, this library is fantastic but lacking of GPU implementation makes it unusable for large production datasets.

**Comment by MattiaVerticchio** [comment]
I think we should also take a look at [RAPIDS cuML](https://github.com/rapidsai/cuml), it’s similar to H2O4GPU, but it’s actively developed and supports more algorithms, including KNN ones. It maintains `sklearn` API design where possible and supports GPU accelerated DataFrames using [cuDF](https://github.com/rapidsai/cudf).

The only issue I can see is that it uses `conda` as the main installation method and doesn’t support `pip`.

**Comment by MattiaVerticchio** [comment]
@pycaret do you think it’s doable to use cuML/cuDF?
They’re only [available](https://rapids.ai/start.html#get-rapids) via Conda or by building from source and I’m not familiar with how `pip` deals with dependencies.

**Comment by Yard1** [comment]
I never tried this, but perhaps it is possible to make conda a dependency of pycaret, and then add a custom build step to use conda to download and install the package? Or perhaps just have the user do it themselves, and throw an exception asking them to do so if they choose to use GPU models without cuML installed.

**Comment by MattiaVerticchio** [comment]
I edited the `compare_models()` function for `pycaret.classification` using the new global `gpu_param`.
Also, I added cuML SVM - radial kernel. Here’s the [notebook](https://www.kaggle.com/mattiavert/pycaret-h2o4gpu-cuml).

These are the improvements in training times:
- `RandomForestClassifier` — 4X
- `XGBClassifier` — 16X
- `GradientBoostingClassifier` — 35X
- `SVC` — 351X

`CatBoostClassifier` and `SGDClassifier` took 3.5X and 3X longer, for now, I’m not including them in the pull request.
I’ll try implementing these methods in `pycaret.regression` too, if it’s OK for you.

**Comment by Yard1** [comment]
@MattiaVerticchio Doss cuML provide their own Catboost wrapper? Or was that with `task_type='GPU' ` for the original Catboost estimator? If not, could you check how that works? I think that for some algorithms, GPU acceleration is helpful only for large datasets.

**Comment by MattiaVerticchio** [comment]
The only routine from cuML as of now is `SVC`. I tried setting `task_type='GPU'` for the official CatBoost estimator.

There is a [CatBoost blog post](https://catboost.ai/docs/concepts/speed-up-training.html) about training speed which advises setting `border_count=32` and I’ll try that too, but of course GPU training could only be worth it beyond a certain dataset size.

**Comment by Yard1** [comment]
Does cuML not provide their own GPU accelerated versions of the estimators? It seems to me to be a good idea to try and cut down on the number of required libraries. Aside from that, looks pretty great. I need to talk it through with @pycaret first but one of the things I want to include in a general refactor is a centralized model repository per module which would make it very easy to change the model classes globally, instead of having to edit the code in every method.

**Comment by MattiaVerticchio** [comment]
Here’s the [API website](https://docs.rapids.ai/api/cuml/stable/api.html). There are a lot of accelerated estimators form Scikit-Learn, but CatBoost is not included in `sklearn` models.

I imported `RandomForestClassifier` and `GradientBoostingClassifier` from `h2o4gpu` because cuML implementation is not stable yet.

**Comment by g0lemXIV** [comment]
Excuse me for interrupting, it will be nice if the user could choose the version of background library. I know from experience GPU acceleration is frustrating on Windows, so it will be nice to make choose before using pycaret to avoid issues and errors.
```bash
pip install pycaret
pip install pycaret[gpu]
``` etc.

**Comment by Yard1** [comment]
We'll have it as a param passed to setup, disabled by default.

**Comment by pycaret** [comment]
@MattiaVerticchio @webpolis 

Hi all,

This issue has been resolved. We have provided the option to use GPU in `pycaret=2.2`. Please check the release notes for more details: https://github.com/pycaret/pycaret/releases

Thanks
